File: provable-proof-js/src/envelope.ts
import { keccak256, sha256 } from '@kuip/provable-sdk';
import type { KayrosData, ProofDataFormat } from './types';

type HashAlgorithm = 'sha256' | 'keccak256';
type JsonObject = Record<string, unknown>;

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

function isRecord(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null;
}

function readPath(root: unknown, path: string[]): unknown {
  let current = root;
  for (const segment of path) {
    if (!isRecord(current) || !(segment in current)) {
      return undefined;
    }
    current = current[segment];
  }
  return current;
}

function firstString(...values: unknown[]): string | undefined {
  for (const value of values) {
    if (typeof value === 'string' && value.length > 0) {
      return value;
    }
  }
  return undefined;
}

function decodeBase64ToBytes(value: string): Uint8Array | undefined {
  const compact = value.trim().replace(/\s+/g, '');
  if (compact.length === 0 || !/^[A-Za-z0-9+/_-]+={0,2}$/.test(compact)) {
    return undefined;
  }
  const normalized = compact.replace(/-/g, '+').replace(/_/g, '/');
  const padded = normalized + '='.repeat((4 - (normalized.length % 4)) % 4);

  try {
    const binary = atob(padded);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  } catch {
    return undefined;
  }
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes).map((byte) => byte.toString(16).padStart(2, '0')).join('');
}

function normalizeHash(value: unknown): string | undefined {
  if (typeof value !== 'string' || value.length === 0) {
    return undefined;
  }
  const hex = value.trim().toLowerCase().replace(/^0x/, '');
  if (/^[0-9a-f]+$/.test(hex) && hex.length % 2 === 0) {
    return hex;
  }
  const bytes = decodeBase64ToBytes(value);
  return bytes ? bytesToHex(bytes) : undefined;
}

function encodeBytesToBase64(bytes: Uint8Array): string {
  let binary = '';
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary);
}

function encodeTextToBase64(value: string): string {
  return encodeBytesToBase64(textEncoder.encode(value));
}

function encodeData(data: unknown): string {
  if (data instanceof Uint8Array) {
    return encodeBytesToBase64(data);
  }
  if (typeof data === 'string') {
    return encodeTextToBase64(data);
  }
  return encodeTextToBase64(JSON.stringify(data));
}

function decodeRequiredBase64(value: string): Uint8Array {
  const bytes = decodeBase64ToBytes(value);
  if (!bytes) {
    throw new Error('Invalid proof data: expected base64 string');
  }
  return bytes;
}

function normalizeHashAlgorithm(value: unknown): HashAlgorithm {
  if (typeof value !== 'string') {
    return 'sha256';
  }
  const normalized = value.toLowerCase().replace(/_/g, '').replace(/-/g, '');
  if (normalized === 'keccak256') {
    return 'keccak256';
  }
  return 'sha256';
}

async function hashBytes(bytes: Uint8Array, algorithm: HashAlgorithm): Promise<string> {
  if (algorithm === 'keccak256') {
    return keccak256(bytes);
  }
  return sha256(bytes);
}

/**
 * Canonical Kayros proof envelope.
 * - `data` is always base64 in serialized proof JSON.
 * - `kayros` is the timestamp metadata returned for that payload hash.
 */
export class KayrosEnvelope<T = unknown> {
  public readonly data: string;
  public readonly data_format: ProofDataFormat;
  public readonly kayros: KayrosData;
  private readonly dataBytes: Uint8Array;

  constructor(data: string, kayros: KayrosData, dataFormat: ProofDataFormat = '') {
    this.data = data.trim();
    this.data_format = dataFormat;
    this.kayros = kayros;
    this.dataBytes = decodeRequiredBase64(this.data);
  }

  static fromData<T>(data: T, kayros: KayrosData, dataFormat: ProofDataFormat = ''): KayrosEnvelope {
    return new KayrosEnvelope(encodeData(data), kayros, dataFormat);
  }

  static fromJSON(jsonText: string): KayrosEnvelope {
    const parsed = JSON.parse(jsonText) as { data?: unknown; data_format?: unknown; kayros?: KayrosData };
    if (!parsed || typeof parsed !== 'object') {
      throw new Error('Invalid proof JSON: expected object');
    }
    if (!('data' in parsed) || !('kayros' in parsed)) {
      throw new Error('Invalid proof JSON: expected { data, kayros }');
    }
    if (typeof parsed.data !== 'string') {
      throw new Error('Invalid proof JSON: expected data to be base64 string');
    }
    if (!isRecord(parsed.kayros)) {
      throw new Error('Invalid proof JSON: expected kayros metadata object');
    }
    const dataFormat = typeof parsed.data_format === 'string'
      ? parsed.data_format as ProofDataFormat
      : '';
    return new KayrosEnvelope(parsed.data, parsed.kayros as KayrosData, dataFormat);
  }

  getDataFormat(): ProofDataFormat {
    return this.data_format || '';
  }

  getDataHash(): string | undefined {
    return normalizeHash(this.kayros.hash);
  }

  getDataType(): string | undefined {
    return firstString(readPath(this.kayros, ['timestamp', 'response', 'data', 'data_type']));
  }

  getDataTypeLabel(): string | undefined {
    return this.getDataType();
  }

  getDataTypeLookupCandidates(): Array<string | undefined> {
    return [this.getDataType()];
  }

  getKayrosHash(): string | undefined {
    return normalizeHash(
      firstString(
        readPath(this.kayros, ['timestamp', 'response', 'response', 'hash']),
        readPath(this.kayros, ['timestamp', 'response', 'data', 'hash_item']),
      ),
    );
  }

  getTimeUUID(): string | undefined {
    return firstString(
      readPath(this.kayros, ['timestamp', 'response', 'response', 'timeuuid']),
      readPath(this.kayros, ['timestamp', 'response', 'data', 'ts']),
    );
  }

  getHashAlgorithm(): string {
    return normalizeHashAlgorithm(this.kayros.hashAlgorithm);
  }

  getData(): Uint8Array {
    return this.dataBytes;
  }

  getDataText(): string {
    return textDecoder.decode(this.dataBytes);
  }

  parseData<U = unknown>(): U {
    return JSON.parse(this.getDataText()) as U;
  }

  toJSON(): { data: string; data_format: ProofDataFormat; kayros: KayrosData } {
    return {
      data: this.data,
      data_format: this.data_format,
      kayros: this.kayros,
    };
  }

  async computeDataHash(): Promise<string> {
    if (this.data_format === 'raw_hash') {
      return bytesToHex(this.dataBytes);
    }
    return hashBytes(this.dataBytes, this.getHashAlgorithm() as HashAlgorithm);
  }
}

File: provable-proof-rs/src/envelope.rs
use crate::types::{KayrosData, KayrosProof, ProofDataFormat};
use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
use base64::Engine;
use provable_sdk::{keccak256, sha256, ProvableError, Result};
use serde::Serialize;

fn decode_base64(value: &str) -> Result<Vec<u8>> {
    BASE64_STANDARD.decode(value.trim()).map_err(|error| {
        ProvableError::new(format!(
            "Invalid proof data: expected base64 string ({})",
            error
        ))
    })
}

fn encode_base64(bytes: &[u8]) -> String {
    BASE64_STANDARD.encode(bytes)
}

fn normalize_hash(value: Option<&str>) -> Option<String> {
    value.and_then(|input| {
        let trimmed = input.trim();
        if trimmed.is_empty() {
            return None;
        }
        let normalized = trimmed.trim_start_matches("0x").to_lowercase();
        if normalized.len() % 2 == 0 && normalized.chars().all(|ch| ch.is_ascii_hexdigit()) {
            return Some(normalized);
        }
        BASE64_STANDARD.decode(trimmed).ok().map(hex::encode)
    })
}

fn first_string(values: &[Option<String>]) -> Option<String> {
    values
        .iter()
        .flatten()
        .find(|value| !value.is_empty())
        .cloned()
}

#[derive(Debug, Clone, PartialEq)]
pub struct KayrosEnvelope {
    pub data: String,
    pub data_format: ProofDataFormat,
    pub kayros: KayrosData,
    data_bytes: Vec<u8>,
}

impl KayrosEnvelope {
    pub fn new(
        data: impl Into<String>,
        kayros: KayrosData,
        data_format: impl Into<String>,
    ) -> Result<Self> {
        let data = data.into();
        let data_bytes = decode_base64(&data)?;
        Ok(Self {
            data,
            data_format: data_format.into(),
            kayros,
            data_bytes,
        })
    }

    pub fn from_bytes(
        data: &[u8],
        kayros: KayrosData,
        data_format: impl Into<String>,
    ) -> Result<Self> {
        Self::new(encode_base64(data), kayros, data_format)
    }

    pub fn from_text(
        data: &str,
        kayros: KayrosData,
        data_format: impl Into<String>,
    ) -> Result<Self> {
        Self::from_bytes(data.as_bytes(), kayros, data_format)
    }

    pub fn from_json_value<T: Serialize>(
        data: &T,
        kayros: KayrosData,
        data_format: impl Into<String>,
    ) -> Result<Self> {
        let serialized = serde_json::to_vec(data)?;
        Self::from_bytes(&serialized, kayros, data_format)
    }

    pub fn from_json_str(json: &str) -> Result<Self> {
        let parsed: KayrosProof = serde_json::from_str(json)?;
        Self::new(
            parsed.data,
            parsed.kayros,
            parsed.data_format.unwrap_or_default(),
        )
    }

    pub fn get_data_format(&self) -> String {
        self.data_format.clone()
    }

    pub fn get_data_hash(&self) -> Option<String> {
        normalize_hash(self.kayros.hash.as_deref())
    }

    pub fn get_data_type(&self) -> Option<String> {
        self.kayros
            .timestamp
            .as_ref()
            .and_then(|timestamp| timestamp.response.data.as_ref())
            .map(|record| record.data_type.clone())
    }

    pub fn get_data_type_label(&self) -> Option<String> {
        self.get_data_type()
    }

    pub fn get_data_type_lookup_candidates(&self) -> Vec<Option<String>> {
        vec![self.get_data_type()]
    }

    pub fn get_kayros_hash(&self) -> Option<String> {
        let direct = self
            .kayros
            .timestamp
            .as_ref()
            .and_then(|timestamp| timestamp.response.response.as_ref())
            .and_then(|response| response.hash.clone());
        let fallback = self
            .kayros
            .timestamp
            .as_ref()
            .and_then(|timestamp| timestamp.response.data.as_ref())
            .map(|record| record.hash_item.clone());
        normalize_hash(first_string(&[direct, fallback]).as_deref())
    }

    pub fn get_timeuuid(&self) -> Option<String> {
        first_string(&[
            self.kayros
                .timestamp
                .as_ref()
                .and_then(|timestamp| timestamp.response.response.as_ref())
                .and_then(|response| response.timeuuid.clone()),
            self.kayros
                .timestamp
                .as_ref()
                .and_then(|timestamp| timestamp.response.data.as_ref())
                .map(|record| record.ts.clone()),
        ])
    }

    pub fn get_hash_algorithm(&self) -> String {
        self.kayros
            .hash_algorithm
            .clone()
            .unwrap_or_else(|| "sha256".to_string())
            .to_lowercase()
            .replace('_', "")
            .replace('-', "")
    }

    pub fn get_data(&self) -> &[u8] {
        &self.data_bytes
    }

    pub fn get_data_text(&self) -> Result<String> {
        String::from_utf8(self.data_bytes.clone())
            .map_err(|error| ProvableError::new(error.to_string()))
    }

    pub fn parse_data_json<T: serde::de::DeserializeOwned>(&self) -> Result<T> {
        Ok(serde_json::from_slice(&self.data_bytes)?)
    }

    pub fn to_proof(&self) -> KayrosProof {
        KayrosProof {
            data: self.data.clone(),
            data_format: Some(self.data_format.clone()),
            kayros: self.kayros.clone(),
        }
    }

    pub fn compute_data_hash(&self) -> Result<String> {
        if self.data_format == "raw_hash" {
            return Ok(hex::encode(&self.data_bytes));
        }
        Ok(match self.get_hash_algorithm().as_str() {
            "keccak256" => keccak256(&self.data_bytes),
            _ => sha256(&self.data_bytes),
        })
    }
}
